cluster-autoscaler/cloudprovider/gce/gce_manager.go

```diff
@@ -522,6 +525,15 @@ func (m *gceManagerImpl) findMigsInRegion(region string, name *regexp.Regexp) ([
 	return links, nil
 }
 
+// GetMigOptions merges default options with user-provided options as specified in the MIG's instance template metadata
+func (m *gceManagerImpl) GetMigOptions(mig Mig, defaults config.NodeGroupAutoscalingOptions) (*config.NodeGroupAutoscalingOptions, error) {
+	template, err := m.migInstanceTemplatesProvider.GetMigInstanceTemplate(mig.GceRef())
+	if err != nil {
+		return nil, err
+	}
+	return m.templates.BuildAutoscalingOptions(template, defaults)
```

Comment:
I don't think this should be called here directly:
 * GetOptions is similar to MinSize() and MaxSize(): CA code calls it at any point in the loop and we generally assume that:
   * The value returned should not change during CA loop. 
   * The call is "free" from performance point of view (ie. no API calls should happen as a result).

The latter is probably already true due to caching in migInstanceTemplatesProvider, but I think the way to update any provider level settings should be to fetch it in Refresh() call and only return cached value here. In this case I propose to move extractAutoscalingOptionsFromKubeEnv and associated logic into forceRefresh() call and only apply the options to defaults in GetMigOptions.

WDYT?

Reply:
Updated as suggested